packages/core/src/ui/UICanvas.ts

```diff
@@ -0,0 +1,608 @@
+    if (preCamera !== val) {
+      this._renderCamera = val;
+      this._updateCameraObserver();
+      this._setRealRenderMode(this._getRealRenderMode());
+    }
+  }
+
+  get resolutionAdaptationStrategy(): ResolutionAdaptationStrategy {
+    return this._resolutionAdaptationStrategy;
+  }
+
+  set resolutionAdaptationStrategy(val: ResolutionAdaptationStrategy) {
+    if (this._resolutionAdaptationStrategy !== val) {
+      this._resolutionAdaptationStrategy = val;
+      const realRenderMode = this._realRenderMode;
+      if (
+        realRenderMode === CanvasRenderMode.ScreenSpaceCamera ||
+        realRenderMode === CanvasRenderMode.ScreenSpaceOverlay
+      ) {
+        this._adapterSizeInScreenSpace();
+      }
+    }
+  }
+
+  get sortOrder(): number {
+    return this._sortOrder;
+  }
+
+  set sortOrder(val: number) {
+    if (this._sortOrder !== val) {
+      this._sortOrder = val;
+      this._realRenderMode === CanvasRenderMode.ScreenSpaceOverlay &&
+        this.scene._componentsManager._overlayCanvasesSortingFlag;
+    }
+  }
+
+  get distance(): number {
+    return this._distance;
+  }
+
+  set distance(val: number) {
+    if (this._distance !== val) {
+      this._distance = val;
+      this._realRenderMode === CanvasRenderMode.ScreenSpaceCamera && this._adapterPoseInScreenSpace();
+    }
+  }
+
+  constructor(entity: Entity) {
+    super(entity);
+    this._componentType = ComponentType.UICanvas;
+    this._transform = <UITransform>entity.transform;
+    this._onEntityModify = this._onEntityModify.bind(this);
+    this._onCanvasSizeListener = this._onCanvasSizeListener.bind(this);
+    this._onCameraModifyListener = this._onCameraModifyListener.bind(this);
+    this._onCameraTransformListener = this._onCameraTransformListener.bind(this);
+    this._onReferenceResolutionChanged = this._onReferenceResolutionChanged.bind(this);
+    // @ts-ignore
+    this._referenceResolution._onValueChanged = this._onReferenceResolutionChanged;
+  }
+
+  raycast(ray: Ray, out: HitResult, distance: number = Number.MAX_SAFE_INTEGER): boolean {
+    const { elements } = this;
+    for (let i = elements.length - 1; i >= 0; i--) {
+      const element = elements[i];
+      if (element._runtimeRaycastEnable && element._raycast(ray, out, distance)) {
+        return true;
+      }
+    }
+    out.entity = null;
+    out.component = null;
+    out.distance = 0;
+    out.point.set(0, 0, 0);
+    out.normal.set(0, 0, 0);
+    return false;
+  }
+
+  /**
+   * @internal
+   */
+  _prepareRender(context: RenderContext): void {
+    const { engine, elements, _realRenderMode: mode } = this;
+    const { enableFrustumCulling, cullingMask, _frustum: frustum } = context.camera;
+    const { frameCount } = engine.time;
+    const renderElement = (this._renderElement = engine._renderElementPool.get());
+    this._updateSortDistance(context.virtualCamera.position);
+    renderElement.set(this.sortOrder, this._sortDistance);
```

Comment:
In 3d renderer , renderElement sets priority and distanceSort generally within the life cycle of _render. Is it appropriate to put it here?